crates/types/src/journal/raw.rs

```diff
@@ -14,111 +14,275 @@ use super::*;
+    entry: Bytes,
 }
 
-impl<H> RawEntry<H> {
-    pub const fn new(header: H, entry: Bytes) -> Self {
+impl<InvokeEnrichmentResult, AwakeableEnrichmentResult>
+    RawEntry<InvokeEnrichmentResult, AwakeableEnrichmentResult>
+{
+    pub const fn new(
+        header: EntryHeader<InvokeEnrichmentResult, AwakeableEnrichmentResult>,
+        entry: Bytes,
+    ) -> Self {
         Self { header, entry }
     }
 
-    pub fn into_inner(self) -> (H, Bytes) {
+    pub fn into_inner(
+        self,
+    ) -> (
+        EntryHeader<InvokeEnrichmentResult, AwakeableEnrichmentResult>,
+        Bytes,
+    ) {
         (self.header, self.entry)
     }
-}
 
-impl<H: EntryHeader> RawEntry<H> {
+    pub fn header(&self) -> &EntryHeader<InvokeEnrichmentResult, AwakeableEnrichmentResult> {
+        &self.header
+    }
+
+    pub fn header_mut(
+        &mut self,
+    ) -> &mut EntryHeader<InvokeEnrichmentResult, AwakeableEnrichmentResult> {
+        &mut self.header
+    }
+
+    pub fn serialized_entry(&self) -> &Bytes {
+        &self.entry
+    }
+
+    pub fn serialized_entry_mut(&mut self) -> &mut Bytes {
+        &mut self.entry
+    }
+
     pub fn ty(&self) -> EntryType {
-        self.header.to_entry_type()
+        self.header.as_entry_type()
+    }
+
+    pub fn map_header<F, TargetInvokeEnrichmentResult, TargetAwakeableEnrichmentResult>(
+        self,
+        mapper: F,
+    ) -> RawEntry<TargetInvokeEnrichmentResult, TargetAwakeableEnrichmentResult>
+    where
+        F: FnOnce(
+            EntryHeader<InvokeEnrichmentResult, AwakeableEnrichmentResult>,
+            &Bytes,
+        )
+            -> EntryHeader<TargetInvokeEnrichmentResult, TargetAwakeableEnrichmentResult>,
+    {
+        let new_header = mapper(self.header, &self.entry);
+        RawEntry {
+            header: new_header,
+            entry: self.entry,
+        }
+    }
+
+    pub fn to_entry<Codec: RawEntryCodec>(self) -> Result<Entry, RawEntryCodecError> {
+        Codec::deserialize(self.ty(), self.entry)
+    }
+
+    pub fn as_entry<Codec: RawEntryCodec>(&self) -> Result<Entry, RawEntryCodecError> {
+        Codec::deserialize(self.ty(), self.entry.clone())
+    }
+
+    pub fn erase_enrichment(self) -> PlainRawEntry {
+        self.map_header(|h, _| h.erase_enrichment())
     }
 }
 
-/// This struct represents headers as they are received from the wire.
 #[derive(Debug, Clone, PartialEq, Eq)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-pub enum RawEntryHeader {
-    PollInputStream { is_completed: bool },
-    OutputStream,
-    GetState { is_completed: bool },
-    SetState,
-    ClearState,
-    Sleep { is_completed: bool },
-    Invoke { is_completed: bool },
-    BackgroundInvoke,
-    Awakeable { is_completed: bool },
-    CompleteAwakeable,
-    Custom { code: u16, requires_ack: bool },
+pub enum EntryHeader<InvokeEnrichmentResult, AwakeableEnrichmentResult> {
```

Comment:
If now every variant has a `requires_ack` field, should this be pulled up?

```
pub enum EntryHeader<InvokeEnrichmentResult, AwakeableEnrichmentResult> {
  requires_ack: bool,
  kind: Kind,
}
```

Reply:
I guess the important question is whether we access the `requires_ack` field when doing the pattern matching on the type of the `EntryHeader`. Maybe this whole problem goes away if we make the `requires_ack` a concept of the invoker (see https://github.com/restatedev/restate/pull/940#discussion_r1422374334).